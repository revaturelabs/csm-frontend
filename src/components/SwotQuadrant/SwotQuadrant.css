.dragDropZone {
    padding: 2rem;
    text-align: center;
    background: lightgrey;
    border-radius: 0.5rem;
    box-shadow: 5px 5px 10px darkgrey;
    height: 500px;
    max-height: 33%;
    overflow-y: scroll;
}
.dragDropZone p {
    color: black
}

.dragDropZone.inside-drag-area {
    opacity: .7;
}