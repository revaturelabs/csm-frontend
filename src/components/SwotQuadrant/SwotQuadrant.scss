@import "../../styles/main";

.dragDropZone {
  padding: 2rem;
  text-align: center;
  background: rgba($revature-light-gray, 0.3);
  border-radius: 0.3rem;
  @include shadow(10px);
  // box-shadow: 5px 5px 10px $revature-dark-gray;
  height: 15rem;
  max-height: 15rem;
  overflow-y: scroll;
  -ms-overflow-style: none;
  scrollbar-width: 3px;
  margin-right: 5px;
  p {
    color: black;
  }
  .inside-drag-area {
    opacity: 0.7;
  }
}

.dragDropZone::-webkit-scrollbar {
  width: 0.3rem;
}

.dragDropZone::-webkit-scrollbar-track {
  border-radius: 0.3rem;
}
 
.dragDropZone::-webkit-scrollbar-thumb {
  border-radius: 0.3rem;
}