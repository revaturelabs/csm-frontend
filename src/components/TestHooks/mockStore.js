export const mockStore = {
	currentAssociate: {},
	categories: [],
	displayCategories: [],
	startDate: new Date(new Date().setDate(new Date().getDate() - 14)),
	endDate: new Date(),
	editable: true,
	moveType: "create",
	batchTopics: [],
	SWOT: {
		date_created: null,
		author: "",
		Strengths: [],
		Weaknesses: [],
		Opportunities: [],
		Threats: [],
		Notes: "",
	},
	dropDepth: 0,
	dropZone: "",
	StrengthsModal: false,
	WeaknessesModal: false,
	OpportunitiesModal: false,
	ThreatsModal: false,
	swotModal: false,
	swotCharts: false,
	currentNote: "",
	currentCategory: "",
	notes: "",
	displaySwots: []
}
