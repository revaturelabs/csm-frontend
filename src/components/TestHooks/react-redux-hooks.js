// Implement hooks here to mock them in a test function.
import { useSelector as originalUseSelector } from "react-redux";
import { useDispatch as originalUseDispatch } from "react-redux";
import { useHistory as originalUseHistory } from "react-redux";
import { useLocation as originalUseLocation } from "react-redux";

export const useSelector = state => originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
export const useHistory = path => originalUseHistory(path);
export const useLocation = () => originalUseLocation();
