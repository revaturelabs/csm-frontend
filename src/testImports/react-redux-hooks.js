/*
		This occludes the information from the Jest library allowing the hooks to
		bypass Jest's requirement for all functions to be mocked. This allows hooks
		to operate properly within the test function and within the tested source
		file. Please instantiate all react-redux hooks used in unit testing within
		this single file. This allows a single import to handle all hooks functions
*/
import {
		useSelector as originalUseSelector,
		useDispatch as originalUseDispatch
} from "react-redux";

export const useSelector = state => originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
