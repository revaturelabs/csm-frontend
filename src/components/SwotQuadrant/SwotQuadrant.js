import React from "react";
import "./SwotQuadrant.scss";
import { useSelector, useDispatch } from "react-redux";
import SwotQuadrantTable from "../SwotQuadrantTable/SwotQuadrantTable";
import SwotNoteModal from "../SwotNoteModal/SwotNoteModal";

/**
 * @param {Object} props The state that has been given to the Swot Quadrant from the Swot Quadrant Table
 * @param {string} props.name This is the name of the section of SWOT Analysis that the Quadrant represents
 */
const SwotQuadrant = (props) => {
  const swotState = useSelector((state) => state.swotReducer);
  const edit = useSelector((state) => state.swotReducer.editable);
  const dispatch = useDispatch();

  /**
   * Tracks that the object being dragged has entered the drop zone
   * @param {Event} event The object being dragged has crossed into the drop zone border, from outside the drop zone,
   *                      OR entered the drop zone from the drop zone border
   * @modifies {number} swotReducer.dropDepth is incremented by 1
   */
  const handleDragEnter = (event) => {
    event.preventDefault();
    event.stopPropagation();
    dispatch({ type: "updateDropDepth", dropDepth: swotState.dropDepth + 1 });
  };

  /**
   * Tracks that the object being dragged has left the drop zone.
   * @param {Event} event The object being dragged has crossed into the drop zone border from the drop zone,
   *                      OR exited the drop zone border into the non-drop zone
   * @modifies {number} swotReducer.dropDepth is decremented by 1
   * @modifies {string} swotReducer.dropZone is cleared out if the dropDepth is less than 1
   */
  const handleDragLeave = (event) => {
    event.preventDefault();
    event.stopPropagation();
    dispatch({ type: "updateDropDepth", dropDepth: swotState.dropDepth - 1 });
    if (swotState.dropDepth > 0) return;
    dispatch({ type: "updateDropZone", dropZone: "" });
  };

  /**
   * Tracks the name of the drop zone the dragged object is currently hovering over
   * @param {Event} event The dragged object is hovering over a drop zone
   */
  const handleDragOver = (event) => {
    event.preventDefault();
    event.stopPropagation();
    if (props.name !== swotState.dropZone) {
      dispatch({ type: "updateDropZone", dropZone: props.name });
    }
  };

  /**
   * Event handler for the dragged object being dropped into the drag zone
   * @param {Event} event The dragged object has been dropped into  the drag zone
   */
  const handleDrop = (event) => {
    event.preventDefault();
    event.stopPropagation();
    const data = event.dataTransfer.getData("text");
    const section = data.split("~")[0];
    const category = data.split("~")[1];
    let type = "create";
    if (edit && section !== props.name) {
      if (section !== "NONE") {
        deleteCategory(section);
        type = "move";
      }
      dispatch({ type: "updateMoveType", move: type });
      createModal(category);
    }
  };

  /**
   * Creates the modal for adding the note to a category.
   * If the category is already present in the current SWOT section, pre-populates the
   * modal's note field with the pre-existing comment, allowing the user to edit the current comment
   * rather than creating a new entry on the SWOT section
   * @param {'string'} category The name of the category that has been dragged and dropped into the drag zone
   */
  const createModal = (category) => {
    const exists = swotState.SWOT[props.name].find(
      (element) => element.category === category
    );
    if (exists !== undefined) {
      dispatch({ type: "updateCurrentNote", note: exists.note });
    }
    showNoteModal(category);
  };

  /**
   * Triggers the modal to appear on screen, after createModal has set the state of the note correctly
   * @param {'string'} category The name of the category that has been dragged and dropped into the drag zone
   */
  const showNoteModal = (category) => {
    dispatch({ type: "updateCategory", category: category });
    dispatch({ type: "toggle" + props.name + "Modal", toggle: true });
  };

  /**
   * Deletes the current category from the SWOT section's category list
   * 1. Copies the appropriate SWOT section's array
   * 2. Gets index of category to be removed
   * 3. Slices array at index
   * 4. Dispatches new array
   */
  const deleteCategory = (section, category) => {
    const index = swotState.SWOT[section].findIndex((elt) => {
      return elt.category === category;
    });
    const _type = "update" + section;
    let new_arr = [...swotState.SWOT[section]];
    new_arr.splice(index, 1);
    dispatch({ type: _type, data: new_arr });
  };

  /**
   *
   * The SWOT section has the section's name, as well as the list of the section's categories
   * The section is a droppable area, including the category items already present in the section
   */
  return (
    <>
      <SwotNoteModal name={props.name}></SwotNoteModal>
      <div
        className={"dragDropZone"}
        id={props.name + "Zone"}
        onDrop={(event) => handleDrop(event)}
        onDragOver={(event) => handleDragOver(event)}
        onDragEnter={(event) => handleDragEnter(event)}
        onDragLeave={(event) => handleDragLeave(event)}
      >
        <h3>{props.name}</h3>
        <SwotQuadrantTable
          name={props.name}
          editHandler={createModal}
          deleteHandler={deleteCategory}
        ></SwotQuadrantTable>
      </div>
    </>
  );
};

export default SwotQuadrant;
