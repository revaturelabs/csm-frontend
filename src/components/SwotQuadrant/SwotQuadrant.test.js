// Component.test.js
import React from "react";
import ReactDOM from "react-dom";
import Enzyme, { shallow, mount, render, fireEvent } from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import { useSelector, useDispatch } from 'react-redux';
import mockStore from "../TestHooks/mockStore.js";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import SwotQuadrant from './SwotQuadrant';

import * as ReactReduxHooks from "../TestHooks/react-redux-hooks";

// It can receive two more parameters, the second one is to specify a factory instead of the jest's automocking feature
jest.mock('react-redux');

describe("Search unit test", () => {
    let wrapper;
    let useEffect;
    let useState;
    let store;
    let dispatch;

    const mockUseState = () => {
        useState.mockImplementationOnce();
    };

    const mockUseEffect = () => {
        useEffect.mockImplementationOnce(f => f());
    };

    const mockUseDispatch = () => {
        useDispatch.mockImplementationOnce();
    };

    beforeEach(() => {
        /* mocking store */
		store = configureStore([thunk])({
			mockStore
		});

	    /* mocking useEffect */
		useEffect = jest.spyOn(React, "useEffect");
		mockUseEffect(); // 2 times
		mockUseEffect(); //

		/* mocking useState */
		useState = jest.spyOn(React, "useState");
        mockUseState();

    	/* mocking useSelector on our mock store */
		jest
	   		.spyOn(ReactReduxHooks, "useSelector")
            .mockImplementation(state => store.getState());

		/* mocking useDispatch on our mock store  */
		jest
	 		.spyOn(ReactReduxHooks, "useDispatch")
	 		.mockImplementation(() => store.dispatch);

		/* mocking useLocation */
		jest
			.spyOn(ReactReduxHooks, "useLocation")
            .mockImplementation((path) => useLocation({
                pathname: path
            }));

        /* mocking useHistory */
        jest
            .spyOn(ReactReduxHooks, "useHistory")
            .mockImplementation(() => useHistory());

        /* shallow rendering */
        wrapper = shallow(<Categories store={store} />);
    });

    describe("on mount", () => {

      	it('Rendering component without props.', () => {
    		const component = shallow(<Categories/>);
    		expect(component).toMatchSnapshot();
      	});

      	it('Rendering component with children.', () => {
    		const component = render(<Categories/>);
    		expect(component).toMatchSnapshot();
      	});

      	it('Mounting, testing list item mapping, and dismounting the component.', () => {
    		const component = mount(<Categories/>);
    		expect(component).toMatchSnapshot();
    		component.unmount();
      	});

    });

});
