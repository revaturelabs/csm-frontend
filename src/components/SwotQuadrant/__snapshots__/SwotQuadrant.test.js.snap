// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwotQuadrant test suite initialization. SwotQuadrant test cases. Rendering component without props. 1`] = `
<Fragment>
  <SwotNoteModal />
  <div
    className="dragDropZone"
    id="undefinedZone"
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <h2 />
    <SwotQuadrantTable
      deleteHandler={[Function]}
      editHandler={[Function]}
    />
  </div>
</Fragment>
`;
